<a href="https://colab.research.google.com/github/Manuelamejiah/Business-intelligence-/blob/main/Reto_1_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

**0. Se carga la Base de datos**

In [ ]:
nxl= '/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB =pd.read_excel(nxl,sheet_name=0)

In [ ]:
# selecciono las variables de interés para el modelo
XDB=XDB[['Age', 'Urban','Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']]

# Y elimino datos faltantes
XDB = XDB.dropna()

In [ ]:
#Determino el número total de datos después de eliminar los valores faltantes
print('Número total de datos después de eliminar valores faltantes:', len(XDB))

Número total de datos después de eliminar valores faltantes: 5948


In [ ]:
#Determino los % de personas que viven en zonas urbanas y zonas rurales.

clasificacion_counts = XDB['Urban'].value_counts()
total_datos = len(XDB)

porcentajes = (clasificacion_counts / total_datos) * 100

print('Porcentaje de datos en cada categoría:\n',porcentajes)



Porcentaje de datos en cada categoría:
 Urban
1.0    79.169469
0.0    20.830531
Name: count, dtype: float64


El 79,17% de las personas analizadas viven en una zona urbana, mientras que el 20,83% viven en zona rural

In [ ]:
print('Rango de Edad:', (XDB['Age'].min(), XDB['Age'].max()))
print('Rango de Altura:', (XDB['Height'].min(), XDB['Height'].max()))
print('Rango de Peso:', (XDB['Weight'].min(), XDB['Weight'].max()))

Rango de Edad: (15, 22)
Rango de Altura: (51.0, 80.0)
Rango de Peso: (72.0, 375.0)


Los intervalos en los que se encuentran las variables de edad, altura y peso son los siguientes:



*   Para la variable Edad tenemos un rango entre los 15 y los 22 años
*   Para la variable Altura tenemos un rango entre las 51 pulgadas y las 80 pulgadas
*   Para la variable peso tenemos un rango entre 72 y 375 libras




In [ ]:
# Especifico mis variables de entrada y salida
XD = XDB[['Age','Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']]
yd = XDB[['Urban']]

**2. Se procede con la implementación del modelo**

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb = GaussianNB()
mnb.fit(XD, yd)

# Obtenemos información del modelo
u = mnb.theta_
print('La media de las variables es:\n', u)
var = mnb.var_
print('La varianza de las variables es:\n', var)
sigma = np.sqrt(var)
print('La desviación estándar de las variables es:\n', sigma)

#Así mismo, identificamos límites superiores e inferiores
LS = u + sigma
print('Límite superior:\n', LS)
LI = u - sigma
print('Límite inferior:\n', LI)


La media de las variables es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La varianza de las variables es:
 [[6.11377773e+00 9.96419474e+00 1.51194652e+01 6.51406931e+00
  1.77789403e+01 1.66650620e+01 9.21039647e+02 1.00161093e+09]
 [6.11792319e+00 1.15645045e+01 1.69373304e+01 5.96364977e+00
  1.76201044e+01 1.76562243e+01 9.01712618e+02 1.30246862e+09]]
La desviación estándar de las variables es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
Límite superior:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
#Evaluamos el modelo con un individuo
XDS = [[34, 12, 15, 2, 24, 69, 160, 48000]]
ydp = mnb.predict(XDS)
print('La predicción del individuo es: \n',ydp)

if ydp==0:
  print('El solicitante vive en área rural')

if ydp==1:
  print('El solicitante vive en área urbana')

La predicción del individuo es: 
 [1.]
El solicitante vive en área urbana


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


A partir del modelo, con las siguientes características:


*   Edad:34
*   Educación de la madre: 12
*   Educación del padre: 15
*   Número de integrantes de la familia: 2
*  Autoestima: 24
*   Altura: 69
*   Peso: 160
*   Igresos: 48000

se determina que el individuo estudiado vive en zona **urbana**









**3. Matriz de confusión**

In [ ]:
from sklearn.metrics import confusion_matrix

#Hacemos un barrido de los datos
ydp2=mnb.predict(XD) #hacemos el pronóstico para todos los datos base
cm=confusion_matrix(yd,ydp2) #crea la matriz de confusión
print("La matriz de confusión es: \n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

#PRIMERA MÉTRICA
#exactitud: como le fue al modelo determinando si vivía en zona rural o urbana
Ex=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud del modelo es: \n",Ex)
#todas las métricas que me den por encima del 75% es bueno.

#tasa de error: mal funcionamiento o error del modelo
TE=(FP+FN)/(VP+VN+FP+FN)
print("La tasa de error del modelo es: \n",TE)

SE=VP/(VP+FN) #Sensibilidad
print('La sensibilidad del modelo es:\n',SE)

SP=VN/(VN+FP) #Especificidad
print('La especificidad del modelo es:\n',SP)

PR=VP/(VP+FP) #Precisión
print('La precisión del modelo es:\n',PR)

PRNeng=VN/(VN+FN) #Precisión Negativa
print('La precisión negativa del modelo es:\n',PRNeng)

La matriz de confusión es: 
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es: 
 0.7908540685944856
La tasa de error del modelo es: 
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La precisión negativa del modelo es:
 0.2727272727272727


A partir de la matriz de confusión se determina que:



1. La exactitud alcanzada por el modelo fue del 79,08%, lo cual representa un buen comportamiento del modelo.

2. La tasa de error alcanzada por el modelo fue de 20,09%


3.  La Sensibilidad alcanzada por el modelo fue del 99,83%, lo que demuestra que hubo una muy buena clasificación de las personas que viven en zonas urbanas
4.  La Especificidad alcanzada por el modelo fue del 0,24%, lo que sugiere que el modelo no clasifica correctamente las personas que viven en zonas rurales




**Análisis de resultados**

De acuerdo con la base de datos se tomaron como referencia para determinar si una persona residía en una zona urbana o rural las variables de Edad,
Educación de la madre,
Educación del padre,
Número de integrantes de la familia,
Autoestima, Altura, Peso e Igresos. Tras una limpieza de datos faltantes, la base de datos quedó con un total de 5948 datos, que se dividían en un 79,17% habitantes de zonas urbanas y 20,83% de zonas rurales.

Al momento de realizar la matriz de confusión para analizar el desempeño del modelo obtenemos resultados muy positivos en los puntos de exactitud, tasa de error y sensibilidad, las cuales obtuvieron unos porcentajes del 79,08%, 20,09% y 99,03% respectivamente, lo que significaría un comportamiento muy aceptable frente a la identificación de quienes viven en zonas rurales. Pero, al evaluar el indicador de especificdad nos damos cuenta que este obtuvo un 0,24% lo que nos dice que al momento de identificar personas que viven en zonas rurales, el modelo falla y no logra identicarlas en su totalidad, haciendo que se tengan malas clasificaciones, y por ende un mal análisis, lo cual terminamos de corroborar con el indicadore de precisión negativa, que nos arroja una precisión del 27,7% al momento de identificar a personas que viven en zonas rurales.

Con esto, podemos decir que el modelo no es muy confiable para la clasificación, principalmente de personas que viven en zonas rurales, y que tal vez es necesario evaluar los datos y modificar el modelo.
